src/layers/basic.jl

```diff
@@ -428,3 +429,50 @@ function Base.show(io::IO, m::Parallel)
   join(io, m.layers, ", ")
   print(io, ")")
 end
+
+"""
+    Embedding(in => out; init=randn)
+
+A lookup table that stores embeddings of dimension `out` 
+for a vocabulary of size `in`. 
+
+This layers is often used to store word embeddings and retrieve them using indices. 
+The input to the layer can be either a vector of indexes
+or the corresponding [onehot encoding](@ref Flux.OneHotArray). 
+
+# Examples
+
+```jldoctest
+julia> m = Embedding(reshape(-6:45, 2, 26) .+ 0.01f0)
+Embedding(26 => 2)
+
+julia> m(5)  # embedding vector for 5th element
+2-element Vector{Float32}:
+ 2.01
+ 3.01
+
+julia> m([6, 15, 15])  # applied to a batch
+2×3 Matrix{Float32}:
+ 4.01  22.01  22.01
+ 5.01  23.01  23.01
+
+julia> ans == m(Flux.OneHotMatrix([6, 15, 15], 26))
+true
+```
+"""
+struct Embedding{W <: AbstractMatrix}
+  weight::W
+end
+
+@functor Embedding
+
+Embedding(dims::Pair{<:Integer, <:Integer}; init = randn32) = Embedding(init(last(dims), first(dims)))
+  
+(m::Embedding)(x::Union{OneHotVector, OneHotMatrix}) = m.weight * x # equivalent to m.weight[:,onecold(x)]
```

Comment:
This is where I think treating `collect(onehot(...))` the same as `onehot()` would be nice. They are both arrays of booleans, so something like this may work:
```suggestion
(m::Embedding)(x::AbstractArray{Bool}) = m.weight * x  # handles OneHotVector, OneHotMatrix
```
(Or it may require a bit more thought which is more specific.)

Reply:
I guess it comes down to whether the `Vector{Bool}` case should be a matmul or index operation.

Reply:
Those should agree when it's actually one-hot (if not OneHot) up to a dropdims:
```
julia> [1 2 3; 4 5 6] * [false, true, false]
2-element Vector{Int64}:
 2
 5

julia> [1 2 3; 4 5 6][:, [false, true, false]]
2×1 Matrix{Int64}:
 2
 5
```
But not when it's not:
```
julia> [1 2 3; 4 5 6] * [false, true, true]
2-element Vector{Int64}:
  5
 11

julia> [1 2 3; 4 5 6][:, [false, true, true]]
2×2 Matrix{Int64}:
 2  3
 5  6
```
Should the latter case be an error? Maybe indexing would be faster although I'm not sure we care about optimising this. Logical indexing can't give you aliasing problems, if I'm thinking correctly.

Reply:
I was mostly speaking from an optimization stand point. If we don't care about this case being fast (it is rare), then the suggested code works for me.

Reply:
Would it be better to change the current onehot paths from 
```julia
function (m::Embedding)(x::Union{OneHotLikeVector{T,L}, OneHotLikeMatrix{T,L,I}}) where {T,L,I}
    size(m.weight, 2) == L || throw(DimensionMismatch("Matrix column must correspond with OneHot size: $(size(m.weight, 2)) != $L"))
  return m(onecold(x))
end
``` 
back to a matmul?

Reply:
Yes definitely, the current implementation is just code duplication.